File: jvm/src/main/java/io/github/juchanei/leetcodeJava/CourseSchedule.java
package io.github.juchanei.leetcodeJava;

import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import java.util.Arrays;
import java.util.HashMap;
import java.util.stream.IntStream;

import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

public class CourseSchedule {
    private int[][] prerequisites;
    private final HashMap<Integer, Boolean> memo = new HashMap<>();

    public boolean canFinish(int numCourses, int[][] prerequisites) {
        this.prerequisites = prerequisites;

        return IntStream.range(0, numCourses)
            .boxed()
            .allMatch(c -> canFinishRecursive(c, numCourses));
    }

    private boolean canFinishRecursive(int course, int left) {
        if (left == 0) return false;
        if (memo.containsKey(course)) return memo.get(course);

        boolean ret = Arrays.stream(prerequisites)
            .filter(pair -> pair[0] == course)
            .map(pair -> pair[1])
            .allMatch(c -> canFinishRecursive(c, left - 1));

        memo.put(course, ret);
        return ret;
    }

    @Nested
    class UnitTest {
        private final CourseSchedule sut = new CourseSchedule();

        @Test
        public void example1() {
            int nc = 2;
            int[][] pre = {{1, 0}};
            assertTrue(sut.canFinish(nc, pre));
        }

        @Test
        public void example2() {
            int nc = 2;
            int[][] pre = {{1, 0}, {0, 1}};
            assertFalse(sut.canFinish(nc, pre));
        }
    }
}

File: jvm/src/main/kotlin/io.github.juchanei.leetcodeKotlin/CourseSchedule.kt
package io.github.juchanei.leetcodeKotlin

import org.assertj.core.api.BDDAssertions.then
import org.junit.jupiter.api.Test

// https://leetcode.com/problems/course-schedule/
class CourseSchedule {
    fun canFinish(numCourses: Int, prerequisites: Array<IntArray>): Boolean {
        val memo = mutableMapOf<Int, Boolean>()

        fun validate(course: Int, left: Int): Boolean =
            when {
                left == 0 -> false
                memo.containsKey(course) -> memo[course]!!
                else -> prerequisites
                    .filter { it[0] == course }
                    .all { validate(it[1], left - 1) }
                    .also { memo[course] = it }
            }

        return (0 until numCourses).all { validate(it, numCourses) }
    }

    @Test
    fun example1() {
        // given
        val nc = 2
        val pre = arrayOf(intArrayOf(1, 0))

        // when
        val ret = canFinish(nc, pre)

        // then
        then(ret).isTrue
    }

    @Test
    fun example2() {
        // given
        val nc = 2
        val pre = arrayOf(intArrayOf(1, 0), intArrayOf(0, 1))

        // when
        val ret = canFinish(nc, pre)

        // then
        then(ret).isFalse
    }
}
